lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
+        for _, parent := range lotMap[name].Parents {
+            if err := visit(parent); err != nil {
+                return err
+            }
+        }
+		// Adding the leaves of the DFS parent tree to the sorted list
+		// guarantees that we'll add the parents before the children
+        sorted = append(sorted, lotMap[name])
+        return nil
+    }
+
+    for name := range lotMap {
+        if err := visit(name); err != nil {
+            return nil, err
+        }
+    }
+
+    return sorted, nil
+}
+
+
+// Initialize the lot configurations based on provided policy, discovered namespaces,
+// and available cache space, handling any necessary merges and validations along the way.
+func initLots(nsAds []server_structs.NamespaceAdV2) ([]Lot, error) {
+	var internalLots []Lot
+
+	policies, err := getPolicyMap()
+	if err != nil {
+		return internalLots, errors.Wrap(err, "unable to parse lotman configuration")
+	}
+
+	// Get the configured policy, which defines any lots we may need to handle
+	// along with merging logic and purge ordering
+	policyName := param.Lotman_EnabledPolicy.GetString()
+	if _, exists := policies[policyName]; !exists {
+		return internalLots, errors.Errorf("enabled policy %s is not defined in the configuration", policyName)
+	}
+	policy := policies[policyName]
+
+	discoverPrefixes := policy.DiscoverPrefixes
+	shouldMerge := policy.MergeLocalWithDiscovered
+	if shouldMerge && !discoverPrefixes {
+		return internalLots, errors.New("MergeLocalWithDiscovered is set to true, but DiscoverPrefixes is set to false. This is not a valid configuration")
+	}
+
+	// policyLotMap will hold the lots defined in the configuration file (if any) provided by the cache admin
+	policyLotMap := make(map[string]Lot)
+	for _, lot := range policy.Lots {
+		policyLotMap[lot.LotName] = lot
+	}
+
+	var lotMap map[string]Lot
+	if discoverPrefixes {
+		directorLotMap, err := configLotsFromFedPrefixes(nsAds)
+		if err != nil {
+			return internalLots, errors.Wrap(err, "error configuring lots from federation prefixes")
+		}
+
+		// Handle potential need to merge discovered namespaces with provided configuration
+		if shouldMerge {
+			log.Debug("Merging lot configuration from discovered namespaces with configured lots")
+			lotMap, err = mergeLotMaps(directorLotMap, policyLotMap)
+			if err != nil {
+				return internalLots, errors.Wrap(err, "error merging discovered namespaces with configured lots")
+			}
+			log.Tracef("Merged lot configuration: %+v", lotMap)
+		} else {
+			lotMap = make(map[string]Lot)
+			// first set things up with the director lots, then overwrite with the policy lots.
+			// This allows cache admins to override any discovered lots with their own configuration.
+			for key, value := range directorLotMap {
+				lotMap[key] = value
+			}
+			for key, value := range policyLotMap {
+				lotMap[key] = value
+			}
+		}
+	} else {
+		lotMap = policyLotMap
+	}
+
+	cacheDisks := param.Cache_DataLocations.GetStringSlice()
```

Comment:
Is it possible that Cache_DataLocations could somehow be nested folders? Have we ensure that isn't the case? Because otherwise this could potentially double count space.

Reply:
This would probably produce a non-functioning cache in the first place. I double checked and xrootd doesn't throw an error if started with nested data disks, but the handling of that shouldn't be in this lotman code. I can add a check higher up in the cache.